sklearn/cluster/hierarchical.py

```diff
@@ -379,6 +379,10 @@ def linkage_tree(X, connectivity=None, n_components=None,
             'Unknown linkage option, linkage should be one '
             'of %s, but %s was given' % (linkage_choices.keys(), linkage))
 
+    if affinity == 'cosine' and np.sum(1 - np.any(X, axis=1)) > 0:
```

Comment:
I hate to be the person always complaining, but "np.sum(1 - np.any(X, axis=1)) > 0" is a strange test. What exactly is it testing? That X is zero everywhere?

Reply:
It makes sure there are no rows with all zeros.
```
[[0,0],
[1,1]] 
```
for example will produce this exception, but the following example won't.
```
[[1,0],
[1,1]] 
```

Reply:
Use `~` instead of `1-` and use np.any instead of `sum > 0`.

Reply:
Then it would be readable as "any rows without any values"

Reply:
I would have written that "np.any(np.all(X == 0, axis=1))". But what is written here is fine.

Thanks!